src/node.cc

```diff
@@ -851,11 +873,21 @@ void PromiseRejectCallback(PromiseRejectMessage message) {
 
   if (event == v8::kPromiseRejectWithNoHandler) {
     callback = env->promise_reject_unhandled_function();
-    value = message.GetValue();
 
+    // Exit the process in case the unhandled promise is garbage collected
+    // without being handled later on.
+    Persistent<Promise>* p = new Persistent<Promise>(env->isolate(), promise);
+    p->SetWeak(p, WeakCallback, WeakCallbackType::kFinalizer);
+    CHECK_EQ(promise->GetAlignedPointerFromInternalField(1), nullptr);
+    promise->SetAlignedPointerInInternalField(1, p);
+
+    value = message.GetValue();
     if (value.IsEmpty())
       value = Undefined(isolate);
   } else if (event == v8::kPromiseHandlerAddedAfterReject) {
+    Persistent<Promise>* p = static_cast<Persistent<Promise>* >(
+        promise->GetAlignedPointerFromInternalField(1));
+    delete p;
```

Comment:
Set the internal field to nullptr here and can you use a constant for the field index?

As well, no space necessary before the last `>` in this glorious C++11 world.